torch/_dynamo/variables/torch.py

```diff
@@ -766,12 +768,30 @@ def speculate_branch(branch):
 
                 guards = state.output.guards
                 nn_modules = state.output.nn_modules
+                side_effects = state.output.side_effects
 
                 # Nub out bits of state that we don't require to be
                 # equal
                 comparable_state = state._replace(
                     output=state.output._replace(
                         guard_state=GuardsCheckpointState(set()),
+                        # New cells are created for all closure variables, including functions in outer scope.
+                        # Then these "side effects" are pruned (see above) based on which of them are accessed.
+                        # Unfortunately this causes different "side effects" per branch.
+                        # We shouldn't care when branches call different sets of functions in outer scope.
+                        side_effects=SideEffects(
+                            id_to_variable={
+                                k: v
+                                for k, v in side_effects.id_to_variable.items()
+                                if not isinstance(v, NewCellVariable)
+                            },
+                            store_attr_mutations={
+                                k: v
+                                for k, v in side_effects.store_attr_mutations.items()
+                                if not isinstance(k, AttributeMutationNew)
+                            },
```

Comment:
So, if this strategy is sound, then I expect we can also get rid of the change I needed in the original PR to prevent allocations from showing up in some cases: https://github.com/pytorch/pytorch/pull/90286/files#diff-6cfc97f79ba69f9eb1fd097057e89f00b8a35797b2b8e683d44ac734d491a68cR176-R185

But that raises some more questions:

* Why in this modified example do you end up with side effects (whereas previously we optimistically assumed no side effects)
* Is it actually sound to drop the allocation side effects? For example, we could imagine that AFTER the conditional, we attempt to write to the cell that maybe was allocated. Now it seems bad if one branch allocated a new cell, whereas the other one didn't. And if it is sound to drop, why do we track the allocations as side effects at all?

@jansel I wonder if you know, as the original author of the side effect code.

Reply:
IIUC the side effects exist whenever there are references to closure variables, but the *pruning* retains different subsets, thus triggering the diff.

Frankly I also considered taking out the pruning. That should also work for fixing this bug, but I'm not sure why it's there.

We could also maybe take the union of allocations from both branches if your second point is an issue...

Reply:
I added the pruning with the idea that if a conditional branch has side effects, but they end up only being visible locally, then it's the same as not having side effects at all (it's not externally observable).

Reply:
That makes sense. Perhaps we need to distinguish new allocations in the branch scope from those in the parent scope and just prune the ones in branch scope?

Reply:
In any case there remains some uneasiness around the case where a new allocation "escapes" the branch, say by returning from a branch. Not sure how to track that.

Reply:
So, maybe it is a bug in pruning? If the allocation is nonlocal (which is the case for cells because we lazily allocate then-- tbh not sure why this is lazy) we maybe shouldn't prune it?

Reply:
OK, so I put up an alternate fix based on this discussion: https://github.com/pytorch/pytorch/pull/92644

WDYT? (I'll land only one of these.)

Reply:
I believe for export (the main use case on `cond`) we basically can't support side effects.  See my comment about setting `self.foo=1` versus `self.foo=2` in the other PR.

All this machinery is for tracking side effects in case the user mutates these objects, but if the user does mutate those objects the export should fail.

So rather than trying to merge side effect structures, I think we can just assert there aren't side effects and drop the structure.

Reply:
@jansel But we can't naively do that, because access to closed over values triggers a new closure cell which counts as a side effect lol

Reply:
We need to track closures in case they get mutated, and we reuse that tracking handle access to them so that post-mutation accesses get the new value.  However, since we don't actually support mutation with export -- we can just copy the structure and let the recursive calls to export them garbage collect them however they want.